// @roots/bud-framework/lib/bud/index.js
import { __decorate } from "tslib";
import { exit } from 'node:process';
import { bootstrap } from '@roots/bud-framework/bootstrap';
import methods from '@roots/bud-framework/methods';
import { bind } from '@roots/bud-support/decorators/bind';
import { BudError, render as renderError } from '@roots/bud-support/errors';
import isFunction from '@roots/bud-support/isFunction';
import isString from '@roots/bud-support/isString';
import isUndefined from '@roots/bud-support/isUndefined';
import logger from '@roots/bud-support/logger';
/**
 * Bud core class
 */
export class Bud {
    /**
     * True when child compilers
     * @readonly
     */
    get hasChildren() {
        return this.children && Object.entries(this.children).length > 0;
    }
    /**
     * True when current instance is a child instance
     * @readonly
     */
    get isChild() {
        return this.root?.context?.label !== this.context?.label;
    }
    /**
     * True when {@link Bud.mode} is `development`
     * @readonly
     */
    get isDevelopment() {
        return this.mode === `development`;
    }
    /**
     * True when {@link Bud.mode} is `production`
     * @readonly
     */
    get isProduction() {
        return this.mode === `production`;
    }
    /**
     * True when current instance is the parent instance
     * @readonly
     */
    get isRoot() {
        return this.root?.context?.label === this.context?.label;
    }
    /**
     * Label
     * @readonly
     */
    get label() {
        return this.context?.label ?? `bud`;
    }
    /**
     * Compilation mode
     *
     * @remarks
     * Either `production` or `development`.
     *
     * @readonly
     * @defaultValue `production`
     */
    get mode() {
        return this.context?.mode ?? `production`;
    }
    /**
     * Promised tasks
     */
    promised = [];
    /**
     * Constructor
     */
    constructor() {
        if (this.implementation)
            this.set(`implementation`, this.constructor);
        Object.entries(methods).map(([prop, value]) => this.set(prop, value));
    }
    /**
     * Boot application services
     */
    async boot() {
        await this.executeServiceCallbacks(`boot`);
    }
    /**
     * Bootstrap the application
     */
    async bootstrap() {
        await this.executeServiceCallbacks(`bootstrap`);
    }
    /**
     * Error handler
     */
    catch(error) {
        renderError(error);
        if (!this.context.ignoreErrors) {
            exit(1);
        }
    }
    /**
     * Execute service callbacks for a given stage
     * @internal
     */
    async executeServiceCallbacks(stage) {
        await this.resolvePromises();
        await this.hooks.fire(stage, this);
    }
    /**
     * Bud initialize
     */
    async initialize(context) {
        if (!context)
            this.catch(`context is required`);
        this.context = { ...(this.context ?? {}), ...context };
        await bootstrap(this);
        return this;
    }
    /**
     * Creates a child and returns the parent instance
     */
    async make(request, setupFn) {
        if (!this.isRoot) {
            this.catch(`bud.make: must be called from the root context`);
        }
        const context = isString(request)
            ? { ...this.context, label: request, root: this }
            : { ...this.context, ...request, root: this };
        if (isUndefined(context.label)) {
            this.catch(`bud.make: context.label must be a string`);
        }
        if (!isUndefined(this.context.filter) &&
            !this.context.filter.includes(context.label)) {
            logger
                .scope(`make`)
                .log(`skipping child instance based on --filter flag:`, context.label);
            return this;
        }
        if (!this.children)
            this.children = {};
        if (this.children[context.label]) {
            this.catch(`bud.make: child instance ${context.label} already exists`);
        }
        logger.scope(`make`).log(`Instantiating ${context.label}`);
        this.children[context.label] = new this.implementation();
        await this.children[context.label].initialize({ ...context });
        if (setupFn) {
            await setupFn(this.children[context.label]);
            await this.children[context.label].resolvePromises();
        }
        if (typeof request !== `string` && request.dependsOn) {
            this.get(context.label)?.hooks.on(`build.dependencies`, request.dependsOn);
        }
        return this;
    }
    /**
     * Await all promised tasks
     */
    promise(promise) {
        this.promised.push(promise);
        return this;
    }
    async resolvePromises() {
        const promises = this.promised;
        this.promised = [];
        await promises.reduce(async (promised, promise) => {
            await promised;
            try {
                await promise(this);
            }
            catch (error) {
                throw error;
            }
        }, Promise.resolve());
    }
    /**
     * Parent {@link Bud} instance
     * @readonly
     */
    get root() {
        return this.context?.root ?? this;
    }
    /**
     * Set a value on the current instance
     * @param key - key
     * @param value - value
     * @param bind - bind value to current instance (default: true, if bindable)
     */
    set(key, input, bind = true) {
        const toBind = bind && isFunction(input) && `bind` in input;
        const value = toBind ? input.bind(this) : input;
        Object.defineProperty(this, key, {
            configurable: true,
            enumerable: true,
            value,
            writable: true,
        });
        logger
            .scope(this.label, `set`)
            .log(`Defined:`, key, `=>`, typeof value);
        return this;
    }
    /**
     * Log message
     * @deprecated
     */
    log(...messages) {
        logger.scope(this.label).log(...messages);
        return this;
    }
    /**
     * Log success
     * @deprecated
     */
    success(...messages) {
        logger.scope(this.label).log(...messages);
        return this;
    }
    /**
     * Log warning
     * @deprecated
     */
    warn(...messages) {
        logger.scope(this.label).warn(...messages);
        return this;
    }
    /**
     * Log info
     * @deprecated
     */
    info(...messages) {
        logger.scope(this.label).info(...messages);
        return this;
    }
    /**
     * Log error
     * @deprecated
     */
    error(error) {
        logger.scope(this.label).error(error);
        return this;
    }
}
__decorate([
    bind
], Bud.prototype, "boot", null);
__decorate([
    bind
], Bud.prototype, "bootstrap", null);
__decorate([
    bind
], Bud.prototype, "catch", null);
__decorate([
    bind
], Bud.prototype, "executeServiceCallbacks", null);
__decorate([
    bind
], Bud.prototype, "initialize", null);
__decorate([
    bind
], Bud.prototype, "make", null);
__decorate([
    bind
], Bud.prototype, "promise", null);
__decorate([
    bind
], Bud.prototype, "resolvePromises", null);
__decorate([
    bind
], Bud.prototype, "set", null);
__decorate([
    bind
], Bud.prototype, "log", null);
__decorate([
    bind
], Bud.prototype, "success", null);
__decorate([
    bind
], Bud.prototype, "warn", null);
__decorate([
    bind
], Bud.prototype, "info", null);
__decorate([
    bind
], Bud.prototype, "error", null);
