// @roots/blade-loader/lib/module-loader/handlers/index.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.extract = exports.make = exports.repository = void 0;
const tslib_1 = require("tslib");
const css = tslib_1.__importStar(require("@roots/blade-loader/module-loader/handlers/css"));
const js = tslib_1.__importStar(require("@roots/blade-loader/module-loader/handlers/js"));
const scss = tslib_1.__importStar(require("@roots/blade-loader/module-loader/handlers/scss"));
const ts = tslib_1.__importStar(require("@roots/blade-loader/module-loader/handlers/ts"));
const vue = tslib_1.__importStar(require("@roots/blade-loader/module-loader/handlers/vue"));
exports.repository = {
    css: { ...css, loader: require.resolve(`./css/index.cjs`) },
    js: { ...js, loader: require.resolve(`./js/index.cjs`) },
    scss: { ...scss, loader: require.resolve(`./scss/index.cjs`) },
    ts: { ...ts, loader: require.resolve(`./ts/index.cjs`) },
    vue: { ...vue, loader: require.resolve(`./vue/index.cjs`) },
};
const make = function ({ extension, loader }) {
    const context = this.context || this.rootContext;
    const request = `${this.resource}.${extension}!=!${loader}!${this.remainingRequest}`;
    return `import ${JSON.stringify(this.utils.contextify(context, request))};`;
};
exports.make = make;
const extract = (source, pattern) => [...source.matchAll(pattern)]
    .map(({ groups }) => groups?.content)
    .filter(Boolean)
    .join(`\n`);
exports.extract = extract;
